banzai/images.py

```diff
@@ -1,272 +1,766 @@
+    def __del__(self):
+        for handle in self._file_handles:
+            handle.close()
+        del self.data
+        del self.mask
 
-    def table_to_hdu(self):
-        table_hdu = fits_utils.table_to_fits(self._data_table)
-        table_hdu.name = self.name
-        return table_hdu
+    @property
+    def extension_name(self):
+        return self.meta.get('EXTNAME').replace('SCI', '')
 
+    @extension_name.setter
+    def extension_name(self, value):
+        self.meta['EXTNAME'] = value
 
-class Image(object):
+    @classmethod
+    def from_fits(cls, hdu: Union[fits.ImageHDU, fits.TableHDU, fits.BinTableHDU]):
+        return cls(hdu.data, hdu.header, name=hdu.header.get('EXTNAME'))
 
-    def __init__(self, runtime_context, filename=None, data=None, data_tables=None,
-                 header=None, extension_headers=None, bpm=None):
-        if header is None:
-            header = fits.Header()
+    @abc.abstractmethod
+    def to_fits(self) -> Union[fits.HDUList, list]:
+        pass
 
-        if data_tables is None:
-            data_tables = {}
 
-        if extension_headers is None:
-            extension_headers = []
+class HeaderOnly(Data):
+    def __init__(self, meta: Union[dict, fits.Header]):
+        super().__init__(data=np.zeros(0), meta=meta)
 
-        if filename is not None:
-            data, header, bpm, extension_headers = fits_utils.open_image(filename)
-            if '.fz' == filename[-3:]:
-                filename = filename[:-3]
-            self.filename = os.path.basename(filename)
+    def to_fits(self):
+        return fits.HDUList([fits.ImageHDU(data=None, header=self.meta)])
 
-        self.data = data
-        self.data_tables = data_tables
-        self.header = header
-        self.bpm = bpm
 
-        self.extension_headers = extension_headers
+class CCDData(Data):
+    def __init__(self, data: Union[np.array, Table], meta: Union[dict, fits.Header],
+                 mask: np.array = None, name: str = '', uncertainty: np.array = None):
+        super().__init__(data=data, meta=meta, mask=mask, name=name)
+        if uncertainty is None:
+            uncertainty = self.read_noise * np.ones(data.shape, dtype=data.dtype) / self.gain
+        self.uncertainty = self._init_array(uncertainty)
 
-        self.request_number = header.get('REQNUM')
-        self.instrument, self.site, self.camera = self._init_instrument_info(runtime_context)
+    def __getitem__(self, section):
+        """
+        Return a new CCDData object with the given section of data
+        :param section: needs to be  in data coords
+        :return:
+        """
+        return self.trim(trim_section=section)
+
+    def __imul__(self, value):
+        self.data *= value
+        self.meta['SATURATE'] *= value
+        self.meta['GAIN'] *= value
+        self.meta['MAXLIN'] *= value
+        return self
+
+    def __idiv__(self, other):
+        self.__imul__(1.0 / other)
+        return self
```

Comment:
I thought __idiv__ and __div__ are replaced with __itruediv__ and __truediv__ in python 3. (https://docs.python.org/3/library/operator.html)

In a little example I did with a simple class, in python 3 I can define custom division using __truediv__ but __div__ does not work because it is depricated. How did you get around this?